### ISSUE TITLE
perf(6482): verify matched FP32/AMP training equivalence and throughput on one pinned GPU fixture

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#7254]
## Parent

Evidence child of #6482. The three implementation slices are already merged:

- DataLoader/pinned-memory path: #6488 / PR #6518 (`156cb30990fafd667ae839e2543b1d75dc448742`);
- opt-in CUDA AMP path: #6509 / PR #6523 (`14b82d452919593145f0545445a4b9da0686cadc`);
- bounded replay eviction: #6508 / PR #6524 (`f216d2f4810b9d2176043f84284053ad5e1464c1`).

This child owns only the remaining measured proof behind the parent criterion “training produces equivalent results (loss curves within noise).”

## Goal

On one pinned predictive-planner training fixture and one declared NVIDIA GPU environment, compare:

1. FP32 with the new loader settings disabled/default;
2. FP32 with the selected pinned-memory/worker settings enabled;
3. AMP with the same loader settings.

Determine whether the optimized paths preserve a bounded short-run training signal and whether they change throughput enough to justify retaining or recommending them. This is training-tooling smoke evidence, not policy-quality or benchmark evidence.

## Frozen comparison contract

Before the AMP arm is inspected, record:

- exact commit, config, dataset/split digest, seed, model, optimizer, batch size, epoch/step budget, device, CUDA/PyTorch versions, loader settings, and warm-up policy;
- the FP32 repeat dispersion used to define the short-run equivalence envelope;
- fixed numerical sanity checks for finite loss, finite gradients, identical example/step counts, checkpoint readability, and manifest completeness;
- throughput measurement boundaries and repetition count.

Keep model, objective, data order, optimizer, batch size, and update budget fixed across arms. AMP and loader settings are the only permitted changes.

## Required outputs

- machine-readable run manifest for every arm;
- step/epoch loss traces and finite-value checks;
- examples/steps processed and checkpoint-load result;
- wall time and examples/steps per second after warm-up;
- peak device-memory observation where available;
- a compact verdict: `equivalent_smoke`, `numerically_divergent`, `unstable_or_nonfinite`, or `inconclusive_environment`.

## Acceptance criteria

- [ ] The exact comparison contract is frozen before AMP outcomes are interpreted.
- [ ] Default/disabled FP32 remains the control and completes with the expected deterministic identity.
- [ ] All arms process identical data identities and update counts.
- [ ] AMP and loader arms produce finite losses/gradients and readable checkpoints.
- [ ] Loss-curve comparison uses the predeclared FP32-repeat envelope; no post-outcome tolerance is introduced.
- [ ] Throughput and memory results include repetitions, warm-up, and environment provenance.
- [ ] Any regression or instability is reported directly; no GPU speedup or policy-quality claim is inferred from one fixture.
- [ ] Parent #6482 receives the exact artifacts and terminal verdict.

## Stop conditions

- Stop if the dataset, ordering, model, optimizer, batch size, or update budget differs between arms.
- Stop on non-finite values, missing manifests, unreadable checkpoints, or unrecorded environment drift.
- Do not tune AMP scaler, batch size, model, objective, or tolerance after seeing comparative outcomes.
- Do not submit a benchmark campaign or promote a planner result from this child.

## Suggested validation

```bash
uv run pytest tests/training -k 'predictive and (amp or loader)' -q
uv run ruff check scripts/training/train_predictive_planner.py tests/training
uv run ruff format --check scripts/training/train_predictive_planner.py tests/training
git diff --check
```

Run the matched short training comparison through the existing config-first predictive-planner command on a GPU-capable host and retain only compact manifests/traces/checksums in Git.

## Evidence boundary

This child can validate implementation-level numerical stability and bounded short-run equivalence on one pinned GPU fixture. It does not establish final policy equivalence, benchmark improvement, general GPU speedup, or dissertation evidence.

**Comment by ll7** [comment]
## Bounded GPU optimization smoke result

The frozen issue #7254 comparison completed on exact commit `b2576e8ad55c50dcfd04aca74ab1bc9de805287b`.

- Host: NVIDIA GeForce RTX 3080, driver 595.71.05, PyTorch `2.13.0+cu130`, CUDA `13.0`.
- Fixture: `issue_7254_predictive_fixture_v1`, SHA-256 `8b8f30ca85a493805277c2b12ae8390dedf60fc9671e8427974e7e2d7bdcd434`, shape `512 x 3 x 4`, horizon 8, seed 7254.
- Arms: FP32 control, FP32 with two loader workers/pinned memory/non-blocking transfer, and AMP with the same loader settings; two repeats per arm.
- Warm-up: first epoch excluded from throughput; the control-repeat dispersion rule and `1e-3` minimum absolute envelope were frozen in the committed YAML before outcomes were inspected.
- All six runs produced finite four-epoch traces, identical example/step counts, and readable checkpoints.

| Arm | Examples/s after warm-up | Peak allocated CUDA bytes | Curve result |
| --- | ---: | ---: | --- |
| FP32 control | 33,712 | 17,415,168 | control |
| FP32 loader | 27,296 | 17,415,168 | equivalent to control |
| AMP + loader | 15,761 | 17,344,000 | equivalent to control |

Classification: `equivalent_smoke`. The optimized arms stayed within the predeclared `1e-3` loss/ADE/FDE envelope, but this tiny fixture measured lower throughput than the single-process FP32 control. That is a diagnostic result, not evidence for a general speedup or a default recommendation; a larger realistic fixture would be needed before making that decision.

The reusable config-first runner and per-epoch timing/peak-memory provenance are in PR `pending` from commit `b2576e8`.

**Comment by ll7** [comment]
## Bounded GPU optimization smoke result (final exact-head rerun)

This final rerun supersedes the earlier handoff and completed on exact commit
`c07d74d38323ef115fd9c123766e0f852a7312e8`.

- Host: NVIDIA GeForce RTX 3080, driver 595.71.05, PyTorch `2.13.0+cu130`, CUDA `13.0`.
- Fixture: `issue_7254_predictive_fixture_v1`, SHA-256 `8b8f30ca85a493805277c2b12ae8390dedf60fc9671e8427974e7e2d7bdcd434`, shape `512 x 3 x 4`, horizon 8, seed 7254.
- Arms: FP32 control, FP32 with two loader workers/pinned memory/non-blocking transfer, and AMP with the same loader settings; two repeats per arm.
- Warm-up: first epoch excluded from throughput; the control-repeat dispersion rule and `1e-3` minimum absolute envelope were frozen in the committed YAML before outcomes were inspected.
- All six runs produced finite four-epoch traces, identical example/step counts, and readable checkpoints.

| Arm | Examples/s after warm-up | Peak allocated CUDA bytes | Curve result |
| --- | ---: | ---: | --- |
| FP32 control | 35,756 | 17,415,168 | control |
| FP32 loader | 26,806 | 17,415,168 | equivalent to control |
| AMP + loader | 21,738 | 17,344,000 | equivalent to control |

Classification: `equivalent_smoke`. The optimized arms stayed within the predeclared `1e-3` loss/ADE/FDE envelope, but this tiny fixture measured lower throughput than the single-process FP32 control. That is a diagnostic result, not evidence for a general speedup or a default recommendation; a larger realistic fixture would be needed before making that decision.

The reusable config-first runner and per-epoch timing/peak-memory provenance are in PR #7263 at
this exact head. The compact output remains worktree-local; this issue comment is the durable
result handoff.

**Comment by ll7** [comment]
## Bounded GPU optimization smoke result (final exact-head rerun)

This final rerun supersedes the earlier handoff and completed on exact commit
`6a395b39819764d98a6c3c7beca4b8bbac7bd1fc`, after refreshing the branch onto live `origin/main`.

- Host: NVIDIA GeForce RTX 3080, driver 595.71.05, PyTorch `2.13.0+cu130`, CUDA `13.0`.
- Fixture: `issue_7254_predictive_fixture_v1`, SHA-256 `8b8f30ca85a493805277c2b12ae8390dedf60fc9671e8427974e7e2d7bdcd434`, shape `512 x 3 x 4`, horizon 8, seed 7254.
- Arms: FP32 control, FP32 with two loader workers/pinned memory/non-blocking transfer, and AMP with the same loader settings; two repeats per arm.
- Warm-up: first epoch excluded from throughput; the control-repeat dispersion rule and `1e-3` minimum absolute envelope were frozen in the committed YAML before outcomes were inspected.
- All six runs produced finite four-epoch traces, identical example/step counts, and readable checkpoints.

| Arm | Examples/s after warm-up | Peak allocated CUDA bytes | Curve result |
| --- | ---: | ---: | --- |
| FP32 control | 33,999 | 17,415,168 | control |
| FP32 loader | 22,818 | 17,415,168 | equivalent to control |
| AMP + loader | 23,246 | 17,344,000 | equivalent to control |

Classification: `equivalent_smoke`. The optimized arms stayed within the predeclared `1e-3` loss/ADE/FDE envelope, but this tiny fixture measured lower throughput than the single-process FP32 control. That is a diagnostic result, not evidence for a general speedup or a default recommendation; a larger realistic fixture would be needed before making that decision.

The reusable config-first runner and per-epoch timing/peak-memory provenance are in PR #7263 at
this exact head. The compact output remains worktree-local; this issue comment is the durable
result handoff.

**Comment by ll7** [comment]
## Bounded GPU optimization smoke result (final exact-head rerun)

This final rerun supersedes the earlier handoff and completed on exact commit
`36abcad38c8bf4e6bec653c07547dffe43c8a4c0`, after refreshing the branch onto live `origin/main`
and adding the fail-closed summary contract.

- Host: NVIDIA GeForce RTX 3080, driver 595.71.05, PyTorch `2.13.0+cu130`, CUDA `13.0`.
- Fixture: `issue_7254_predictive_fixture_v1`, SHA-256 `8b8f30ca85a493805277c2b12ae8390dedf60fc9671e8427974e7e2d7bdcd434`, shape `512 x 3 x 4`, horizon 8, seed 7254.
- Arms: FP32 control, FP32 with two loader workers/pinned memory/non-blocking transfer, and AMP with the same loader settings; two repeats per arm.
- Warm-up: first epoch excluded from throughput; the control-repeat dispersion rule and `1e-3` minimum absolute envelope were frozen in the committed YAML before outcomes were inspected.
- All six runs produced finite four-epoch traces, identical example/step counts, and readable checkpoints.

| Arm | Examples/s after warm-up | Peak allocated CUDA bytes | Curve result |
| --- | ---: | ---: | --- |
| FP32 control | 32,842 | 17,415,168 | control |
| FP32 loader | 23,109 | 17,415,168 | equivalent to control |
| AMP + loader | 18,960 | 17,344,000 | equivalent to control |

Classification: `equivalent_smoke`. The optimized arms stayed within the predeclared `1e-3` loss/ADE/FDE envelope, but this tiny fixture measured lower throughput than the single-process FP32 control. That is a diagnostic result, not evidence for a general speedup or a default recommendation; a larger realistic fixture would be needed before making that decision.

The reusable config-first runner and per-epoch timing/peak-memory provenance are in PR #7263 at
this exact head. The compact output remains worktree-local; this issue comment is the durable
result handoff.

**Comment by ll7** [comment]
## Bounded GPU optimization smoke result (final exact-head rerun)

This final rerun supersedes the earlier handoff and completed on exact commit
`33e1aabbb778a3ef4cbe9dc379c0b3ced990fc9a`, after refreshing the branch onto live `origin/main`
and adding the fail-closed summary contract.

- Host: NVIDIA GeForce RTX 3080, driver 595.71.05, PyTorch `2.13.0+cu130`, CUDA `13.0`.
- Fixture: `issue_7254_predictive_fixture_v1`, SHA-256 `8b8f30ca85a493805277c2b12ae8390dedf60fc9671e8427974e7e2d7bdcd434`, shape `512 x 3 x 4`, horizon 8, seed 7254.
- Arms: FP32 control, FP32 with two loader workers/pinned memory/non-blocking transfer, and AMP with the same loader settings; two repeats per arm.
- Warm-up: first epoch excluded from throughput; the control-repeat dispersion rule and `1e-3` minimum absolute envelope were frozen in the committed YAML before outcomes were inspected.
- All six runs produced finite four-epoch traces, identical example/step counts, and readable checkpoints.

| Arm | Examples/s after warm-up | Peak allocated CUDA bytes | Curve result |
| --- | ---: | ---: | --- |
| FP32 control | 32,500 | 17,415,168 | control |
| FP32 loader | 24,106 | 17,415,168 | equivalent to control |
| AMP + loader | 18,649 | 17,344,000 | equivalent to control |

Classification: `equivalent_smoke`. The optimized arms stayed within the predeclared `1e-3` loss/ADE/FDE envelope, but this tiny fixture measured lower throughput than the single-process FP32 control. That is a diagnostic result, not evidence for a general speedup or a default recommendation; a larger realistic fixture would be needed before making that decision.

The reusable config-first runner and per-epoch timing/peak-memory provenance are in PR #7263 at
this exact head. The compact output remains worktree-local; this issue comment is the durable
result handoff.

**Comment by ll7** [comment]
## Corrective evidence handoff

Issue #7254 is reopened for review of replacement [PR #7364](https://github.com/ll7/robot_sf_ll7/pull/7364), whose exact head is `5bd1f0d29fb19592bd47d61bda487670ee6438fc`.

The durable compact manifest is [`docs/context/evidence/issue_7254_predictive_optimization_smoke_2026-08-17/summary.json`](https://github.com/ll7/robot_sf_ll7/blob/5bd1f0d29fb19592bd47d61bda487670ee6438fc/docs/context/evidence/issue_7254_predictive_optimization_smoke_2026-08-17/summary.json). It binds the corrected smoke to runtime commit `74a55c43f46edca6d1d928074a8a890e608065af`, dataset SHA-256 `8b8f30ca…7bdcd434`, split SHA-256 `a037d8e1…e2a9019`, four identical train-order digests, one environment fingerprint, identical update counts, and six strict checkpoint loads.

The frozen-envelope result is `equivalent_smoke` for the bounded diagnostic fixture. Fresh throughput was 33,678.22 examples/s for FP32 control, 21,239.02 for FP32 loader, and 22,104.56 for AMP+loader. Both optimized arms were slower in this run; this supports no default-performance, benchmark, policy, or paper claim.

The issue should remain open until independent review and hosted checks validate the corrective PR and acceptance boxes are explicitly reconciled.

**Comment by ll7** [comment]
## State synchronization — 2026-08-19

The GPU smoke is complete; the active work is corrective PR #7364, currently open and draft at head `74c9856d8c61d17a58191d8f3284ed2f33c90303`. The stale `state:blocked-no-code-slice` qualifier was removed and `state:working` retained with `needs-review`.

Do not repeat the GPU comparison unless #7364 changes the matched-order or evidence-producing code and the frozen contract requires an exact-head refresh. The current exit gate is fresh current-base validation, evidence/checksum verification, independent exact-head review, and hosted CI. After merge, reconcile the acceptance boxes and close this issue with the bounded `equivalent_smoke` verdict; do not infer a default-performance recommendation or general speedup.

**Comment by ll7** [comment]
## Completion receipt — 2026-08-20

PR #7263 merged at exact head `33e1aabbb778a3ef4cbe9dc379c0b3ced990fc9a` with merge commit `481711033e3b959cbc12a1a3e96bd1cc9e4f2d02`.

The frozen six-run GPU smoke completed with finite traces, identical data and update identities, readable checkpoints, and the terminal classification `equivalent_smoke`. The optimized arms remained inside the predeclared curve envelope, while the small fixture did not show a throughput gain over the FP32 control.

The child contract is complete. Parent #6482 retains the bounded result. Any larger-fixture qualification or default-performance decision requires separate scope. This closure does not establish policy equivalence, benchmark improvement, general GPU speedup, planner quality, or dissertation evidence.
